plugins/inputs/turbostat/turbostat.go

```diff
@@ -0,0 +1,166 @@
+//go:generate ../../../tools/readme_config_includer/generator
+//go:build linux && amd64
+
+package turbostat
+
+import (
+	"bufio"
+	"context"
+	_ "embed"
+	"errors"
+	"fmt"
+	"io"
+	"os/exec"
+	"sync"
+	"time"
+
+	"github.com/jpillora/backoff"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/inputs"
+	"github.com/influxdata/telegraf/plugins/inputs/turbostat/parser"
+)
+
+//go:embed sample.conf
+var sampleConfig string
+
+type Turbostat struct {
+	Command []string        `toml:"command"`
+	Log     telegraf.Logger `toml:"-"`
+
+	cancel context.CancelFunc
+}
+
+func (*Turbostat) Gather(telegraf.Accumulator) error {
+	return nil
+}
+
+func (*Turbostat) SampleConfig() string {
+	return sampleConfig
+}
+
+// Start starts the plugin.
+// Maintains a Turbostat process and parses metrics from stdout.
+func (t *Turbostat) Start(a telegraf.Accumulator) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	t.cancel = cancel
+	go func() {
+		retryDelay := &backoff.Backoff{
+			Min:    1 * time.Second,
+			Max:    30 * time.Second,
+			Factor: 2,
+			Jitter: false,
+		}
+		for {
+			t.Log.Info("Starting Turbostat")
+			startedAt := time.Now()
+			err := t.runTurbostat(ctx, a)
+			t.Log.Errorf("Turbostat encountered an error: %s", err)
+			// Exit the goroutine if the context was cancelled.
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+			// If Turbostat ran for more than 30s, there likely was
+			// no major issue and we reset the backoff to the minimum.
+			if time.Since(startedAt) > 30*time.Second {
+				retryDelay.Reset()
+			}
+			delay := retryDelay.Duration()
+			t.Log.Infof("Restarting Turbostat in %.1fs", delay.Seconds())
+			// Wait until the delay elapses or the context is cancelled.
+			select {
+			case <-ctx.Done():
+				return
+			case <-time.After(delay):
+			}
+		}
+	}()
+	return nil
+}
+
+// Stop stops the plugin.
+func (t *Turbostat) Stop() {
+	t.cancel()
+}
+
+func init() {
+	inputs.Add("turbostat", func() telegraf.Input {
+		return &Turbostat{}
+	})
+}
+
+// Reads error lines from Turbostat stderr and logs them.
+func processStderr(r io.Reader, log telegraf.Logger) error {
+	s := bufio.NewScanner(r)
+	for s.Scan() {
+		line := s.Text()
+		log.Info(line)
+	}
+	return s.Err()
+}
+
+// Starts a Turbostat process. One goroutine parses stdout for metrics and
+// writes them to the accumulator. Another goroutine monitors stderr for
+// messages and logs them. If an error occurs, the process and all goroutines
+// terminate, and the function returns the first error encountered. Since
+// Turbostat is not expected to terminate, this function never returns nil.
+func (t *Turbostat) runTurbostat(c context.Context, a telegraf.Accumulator) error {
+	if len(t.Command) < 1 {
+		return errors.New("no command provided")
+	}
+	ctx, cancel := context.WithCancel(c)
+	defer cancel()
```

Comment:
Why are you creating another context?

Reply:
It's for cancellation. I am now using internal/process in https://github.com/influxdata/telegraf/pull/17368 which takes care of this, so we no longer explicitly create a new context in the plugin.